search/includes/classes/class-versioning.php

```diff
@@ -316,4 +359,168 @@ public function activate_version( Indexable $indexable, int $version_number ) {
+	 * _other_ index versions to keep them in sync
+	 * 
+	 * @param $queued_objects Multidimensional array of queued objects, keyed first by type, then index version
+	 */
+	public function replicate_queued_objects_to_other_versions( $queued_objects ) {
+		if ( ! is_array( $queued_objects ) || empty( $queued_objects ) ) {
+			return;
+		}
+
+		// Loop over every type of object that was changed
+		foreach ( $queued_objects as $object_type => $objects_by_version ) {
+			$indexable = \ElasticPress\Indexables::factory()->get( $object_type );
+
+			// If it's not a valid indexable, just skip
+			if ( ! $indexable ) {
+				continue;
+			}
+
+			$versions = $this->get_versions( $indexable );
+
+			// Do we have any other index versions for this type? If not, nothing to do.
+			if ( ! $versions || empty( $versions ) ) {
+				continue;
+			}
+
+			$active_version_number = $this->get_active_version_number( $indexable );
+
+			// Were there any changes to the active version? If not, we skip - we don't keep replicate non-active indexes to others
+			if ( ! isset( $objects_by_version[ $active_version_number ] ) || empty( $objects_by_version[ $active_version_number ] ) ) {
+				continue;
+			}
+
+			// Other index versions, besides active
+			$inactive_versions = $this->get_inactive_versions( $indexable );
+
+			// There were changes for active version - now we need to loop over every object that was queued for the active version and replicate that job to the other versions
+			foreach ( $inactive_versions as $version ) {
+				$this->set_current_version_number( $indexable, $version['number'] );
+
+				foreach ( $objects_by_version[ $active_version_number ] as $entry ) {
+					$object_id = $entry['object_id'];
+					$options = is_array( $entry['options'] ) ? $entry['options'] : array();
+
+					// Override the index version in the options
+					$options['index_version'] = $version['number'];
+
+					\Automattic\VIP\Search\Search::instance()->queue->queue_object( $object_id, $object_type, $options );
+				}
+
+				$this->reset_current_version_number( $indexable );
+			}
+		}
+	}
+
+	public function filter__pre_ep_index_sync_queue( $bail, $sync_manager, $indexable_slug ) {
+		$indexable = \ElasticPress\Indexables::factory()->get( $indexable_slug );
+
+		// If it's not a valid indexable, just skip
+		if ( ! $indexable ) {
+			return $bail;
+		}
+
+		$inactive_versions = $this->get_inactive_versions( $indexable );
+
+		// If there are no inactive versions or nothing in the queue, we can just skip
+		if ( empty( $inactive_versions ) || empty( $sync_manager->sync_queue ) ) {
+			return $bail;
+		}
+
+		foreach ( $inactive_versions as $version ) {
+			foreach ( $sync_manager->sync_queue as $object_id ) {
+				$options = array(
+					'index_version' => $version['number'],
+				);
+
+				\Automattic\VIP\Search\Search::instance()->queue->queue_object( $object_id, $indexable_slug, $options );
```

Comment:
Here I have a similar comment to the one about the SyncManager: we are writing entries to the queue table before the sync operation to the active index is carried out. So if this write fails for some reason, the object won't be synced to the active index, and this is something we should prioritize.

Maybe we could also add these to the queued_objects array so they are all processed together in `action__shutdown` (provided the correct priority is set to guarantee it's executed after this filter).

Maybe I'm being too conservative here, and there shouldn't be any real problem to write queued operations to the DB before sending the synchronous ones to ES, but I just leave it here as an option to consider :)

Reply:
I think we're good here now that we moved our shutdown hook to priority `100` - we'll always run after EP does its thing (unless EP changed priorities for some reason).

Reply:
The shutdown hook affects only the queued objects, but this code is enqueueing the objects that are processed synchronously. And this filter (pre_ep_index_sync_queue) is applied before the actual sync to ES happens (look at SyncManager::index_sync_queue).

Maybe I'm wrong with the workflows. What I've understood is:

* An object needs to be indexed
** Option 1: it's indexed synchronously (by the SyncManager). We enqueue it for other index versions by adding the ep_index_sync_queue filter, which is processed by this method. The object is enqueued immediately during the filter execution.
** Option 2: it's offloaded to the queue. When it's enqueued, the new action vip_search_indexing_object_queued is called. It's then added to the queued_objects array in this class, and effectively enqueued on the shutdown action.

Is this correct?

Anyway, it shouldn't be a big deal. The shutdown hook was more worrying as it's normally affecting a much larger amount of objects to reindex. But when objects are indexed synchronously there are few of them. So the chance of failure is lower (I expect).

Reply:
Yep that's correct.